lib/puppet/pops/types/types.rb

```diff
@@ -666,12 +677,21 @@ def _assignable?(o, guard)
 # @api public
 #
 class PScalarType < PAnyType
+  KNOWN_SCALARS = [String, Integer, Float, TrueClass, FalseClass, Regexp, Time::Timestamp, Time::Timespan, SemanticPuppet::Version, SemanticPuppet::VersionRange]
+  KNOWN_NON_SCALARS = [Array, Hash, PAnyType, NilClass]
+
   def self.register_ptype(loader, ir)
     create_ptype(loader, ir, 'AnyType')
   end
 
   def instance?(o, guard = nil)
-    assignable?(TypeCalculator.infer(o), guard)
+    if KNOWN_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      true
+    elsif KNOWN_NON_SCALARS.any? { |scalar_class| o.is_a?(scalar_class) }
+      false
```

Comment:
This can be improved for sure. There are two ways to do that:
1. Replace Integer, Float, Timespan, and Timestamp with just Numeric. They are all subclasses anyway so the #is_a? would cover them.
2. Do what you suggest and use a Hash. That will however require that the Hash contains all possible subclasses.
Given that the list will be 3 entries shorter if I change according to #1, I think that would be my preference. Do you agree?